### DISCUSSION TITLE
Predicting stability for mutations of wildtype

### DISCUSSION THREAD

**Discussion by blazespinnaker** [facebookresearch/esm#293] [Q&A] [answered]
Apologies in advance for the sophomoric nature of this question.

Kaggle is hosting a contest for predicting thermodynamic stability 

https://www.kaggle.com/competitions/novozymes-enzyme-stability-prediction

You can see a description of the train and test data here:

https://www.kaggle.com/competitions/novozymes-enzyme-stability-prediction/data

>For the training set, the protein stability (experimental melting temperature) data includes natural sequences, as well as engineered sequences with single or multiple mutations upon the natural sequences. The data are mainly from different sources of published studies such as [Meltome atlas—thermal proteome stability across the tree of life](https://doi.org/10.1038/s41592-020-0801-4). Many other public datasets exist for protein stability; please see the competition [Rule 7C]

>(https://www.kaggle.com/competitions/novozymes-enzyme-stability-prediction/rules) for external data usage requirements. There are also other publicly available methods which can predict protein stabilities such as [ESM](https://doi.org/10.1073/pnas.2016239118), [EVE]
>(https://doi.org/10.1038/s41586-021-04043-8) and [Rosetta](https://doi.org/10.1016/B978-0-12-381270-4.00019-6) etc., without using the provided training set. These methods may also be used as part of the competition.

>The test set contains experimental melting temperature of over 2,413 single-mutation variant of an enzyme (GenBank: KOC15878.1), obtained by Novozymes A/S. The amino acid sequence of the wild type is:

>VPVNPEPDATSVENVALKTGSGDSQSDPIKADLEVKGQSALPFDVDCWAILCKGAPNVLQRVNEKTKNSNRDRSGANKGPFKDPQKWGIKALPPKNPSWSAQDFKSPEEYAFASSLQGGTNAILAPVNLASQNSQGGVLNGFYSANKVAQFDPSKPQQTKGTWFQITKFTGAAGPYCKALGSNDKSVCDKNKNIAGDWGFDPAKWAYQYDEKNNKFNYVGK

I've looked through the codebase, the one shot variant prediction script in particular, but I am at a loss on where to start.

The contest holders recommend ESM as method to perform these predictions, but I am unclear on how to do so.  I've read through the papers, but I'm still unsure.  Any suggestions on where to start?  Feel free to respond here or on Kaggle itself.  I'll make public any code I write as a result of the answer.

Cheers

**Comment by tomsercu** [comment, marked as answer]
Take a look at the `label_row()` function of [predict.py](https://github.com/facebookresearch/esm/blob/main/examples/variant-prediction/predict.py#L107-L115).
It implements the difference of probability of mutated (`mt`) with wildtype (`wt`) sequences, one of the strategies described in the paper.

**Comment by blazespinnaker** [comment]
Curious, I get fairly poor results with the esm1 models.  (around -0.05, spearman), but 0.17 with esm2.   It was 0.234 spearman with the esm2 using the attention contacts from the quick start code.

I did get a warning though, when loading the esm1 models - is this expected?

```
Downloading: "https://dl.fbaipublicfiles.com/fair-esm/models/esm1v_t33_650M_UR90S_1.pt" to /root/.cache/torch/hub/checkpoints/esm1v_t33_650M_UR90S_1.pt
/root/.cache/torch/hub/facebookresearch_esm_main/esm/pretrained.py:216: UserWarning: Regression weights not found, predicting contacts will not produce correct results.
  "Regression weights not found, predicting contacts will not produce correct results."
```
